# Kolmogrov Smirnov normality test
###### written by Mario

Data ini merupakan data penduduk pada desa Karanganyar yang diambil menggunakan metode random sampling.<br> Pada notebook ini akan dilakukan percobaan pengujian statistik dengan hipotesis apakah data berdistribusi normal, dengan <br>__H0 = Data berdistribusi normal & H1 = Data tidak berdistribusi normal__

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import collections

C:\Users\Mario\anaconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Mario\anaconda3\lib\site-packages\numpy\.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll
C:\Users\Mario\anaconda3\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


In [2]:
data = pd.read_csv('path to file', index_col='Nama Anak')

In [3]:
data = data['Penghasilan Ayah'].head(51)

In [4]:
data = data.drop('Rizal C')

In [5]:
data = pd.DataFrame(data)

### Menghitung nilai frekuensi kumulatif dan frekuensi proporsi

In [6]:
#sorting the data
data = data.sort_values(['Penghasilan Ayah'])

In [7]:
counter = dict(collections.Counter(data['Penghasilan Ayah']))

In [8]:
#get the unique data
data_baru = pd.DataFrame(counter.items(), columns=['Penghasilan','frequency'])
data = data_baru

In [9]:
#membuat column frequency cumulative
data['f kumulatif'] = data['frequency'].cumsum()

In [10]:
#membuat column kumulatif proporsi dengan rumus f kumulatif / n
data['kp'] = data['f kumulatif']/50

In [11]:
#membuat column Z dengan rumus (Xi - mean) / sd
data['Z'] = (data['Penghasilan'] - 3119932.48)/data['Penghasilan'].std()
data['Z'] = data['Z'].round(2)

In [12]:
data

Penghasilan  frequency  f kumulatif    kp     Z
0        150000          1            1  0.02 -1.44
1        300000          1            2  0.04 -1.37
2        500000          2            4  0.08 -1.27
3        750000          1            5  0.10 -1.15
4        900000          1            6  0.12 -1.08
5       1000000          5           11  0.22 -1.03
6       1500000          1           12  0.24 -0.79
7       2000000          9           21  0.42 -0.54
8       2500000          1           22  0.44 -0.30
9       3000000          6           28  0.56 -0.06
10      3800000          2           30  0.60  0.33
11      4000000          3           33  0.66  0.43
12      4500000          1           34  0.68  0.67
13      4798312          2           36  0.72  0.81
14      5000000         10           46  0.92  0.91
15      6000000          3           49  0.98  1.40
16      6200000          1           50  1.00  1.49

### Perhitungan table kolom 'z table' mengacu pada table di bawah dengan nilai absolut dari kolom 'z'

<img style="display: block;-webkit-user-select: none;margin: auto;cursor: zoom-out;background-color: hsl(0, 0%, 90%);transition: background-color 300ms;" src="img/z-score.png" width="600" height="812">

In [13]:
#membuat column table Z
tampungan_z_table = np.array([0.4251,0.4147,0.3980,0.3749,0.3599,0.3485,0.2852,0.2054,0.1179,
                             0.0239,0.1293,0.1664,0.2486,0.2910,0.3186,0.4192,0.4319])
data['Z table'] = tampungan_z_table

In [14]:
data

Penghasilan  frequency  f kumulatif    kp     Z  Z table
0        150000          1            1  0.02 -1.44   0.4251
1        300000          1            2  0.04 -1.37   0.4147
2        500000          2            4  0.08 -1.27   0.3980
3        750000          1            5  0.10 -1.15   0.3749
4        900000          1            6  0.12 -1.08   0.3599
5       1000000          5           11  0.22 -1.03   0.3485
6       1500000          1           12  0.24 -0.79   0.2852
7       2000000          9           21  0.42 -0.54   0.2054
8       2500000          1           22  0.44 -0.30   0.1179
9       3000000          6           28  0.56 -0.06   0.0239
10      3800000          2           30  0.60  0.33   0.1293
11      4000000          3           33  0.66  0.43   0.1664
12      4500000          1           34  0.68  0.67   0.2486
13      4798312          2           36  0.72  0.81   0.2910
14      5000000         10           46  0.92  0.91   0.3186
15      6000000          3           49  0.98  1.40   0.4192
16      6200000          1           50  1.00  1.49   0.4319

### Perhitungan table kolom 'f(z)' mengacu pada table di bawah dengan nilai acuan dari kolom 'z'

<img style="display: block;-webkit-user-select: none;margin: auto;cursor: zoom-out;background-color: hsl(0, 0%, 90%);transition: background-color 300ms;" src="https://i0.wp.com/zoebeesley.com/wp-content/uploads/2018/11/z-table1.png?resize=628%2C840&amp;ssl=1" width="600" height="812">

In [15]:
#membuat column fungsi z berdasarkan table z
tampungan_fungsi_z = np.array([0.07493,0.08534,0.10204,0.12507,0.14007,0.15151,0.21476,0.29460,0.38209,0.47608,0.62930,
                              0.66640,0.74857,0.79103,0.81859,0.91924,0.93189])
data["f(z)"] = tampungan_fungsi_z

In [16]:
data

Penghasilan  frequency  f kumulatif    kp     Z  Z table     f(z)
0        150000          1            1  0.02 -1.44   0.4251  0.07493
1        300000          1            2  0.04 -1.37   0.4147  0.08534
2        500000          2            4  0.08 -1.27   0.3980  0.10204
3        750000          1            5  0.10 -1.15   0.3749  0.12507
4        900000          1            6  0.12 -1.08   0.3599  0.14007
5       1000000          5           11  0.22 -1.03   0.3485  0.15151
6       1500000          1           12  0.24 -0.79   0.2852  0.21476
7       2000000          9           21  0.42 -0.54   0.2054  0.29460
8       2500000          1           22  0.44 -0.30   0.1179  0.38209
9       3000000          6           28  0.56 -0.06   0.0239  0.47608
10      3800000          2           30  0.60  0.33   0.1293  0.62930
11      4000000          3           33  0.66  0.43   0.1664  0.66640
12      4500000          1           34  0.68  0.67   0.2486  0.74857
13      4798312          2           36  0.72  0.81   0.2910  0.79103
14      5000000         10           46  0.92  0.91   0.3186  0.81859
15      6000000          3           49  0.98  1.40   0.4192  0.91924
16      6200000          1           50  1.00  1.49   0.4319  0.93189

### Menentukan nilai a1 dan a2

$a2 = |kp-f(z)|$ <br> $a1 = |a2 - (f/n)|$

In [17]:
data['a2'] = abs(data['kp'] - data["f(z)"])
data['a1'] = abs(data['a2'] - (data["frequency"]/50))

In [18]:
data

Penghasilan  frequency  f kumulatif    kp     Z  Z table     f(z)  \
0        150000          1            1  0.02 -1.44   0.4251  0.07493   
1        300000          1            2  0.04 -1.37   0.4147  0.08534   
2        500000          2            4  0.08 -1.27   0.3980  0.10204   
3        750000          1            5  0.10 -1.15   0.3749  0.12507   
4        900000          1            6  0.12 -1.08   0.3599  0.14007   
5       1000000          5           11  0.22 -1.03   0.3485  0.15151   
6       1500000          1           12  0.24 -0.79   0.2852  0.21476   
7       2000000          9           21  0.42 -0.54   0.2054  0.29460   
8       2500000          1           22  0.44 -0.30   0.1179  0.38209   
9       3000000          6           28  0.56 -0.06   0.0239  0.47608   
10      3800000          2           30  0.60  0.33   0.1293  0.62930   
11      4000000          3           33  0.66  0.43   0.1664  0.66640   
12      4500000          1           34  0.68  0.67   0.2486  0.74857   
13      4798312          2           36  0.72  0.81   0.2910  0.79103   
14      5000000         10           46  0.92  0.91   0.3186  0.81859   
15      6000000          3           49  0.98  1.40   0.4192  0.91924   
16      6200000          1           50  1.00  1.49   0.4319  0.93189   

         a2       a1  
0   0.05493  0.03493  
1   0.04534  0.02534  
2   0.02204  0.01796  
3   0.02507  0.00507  
4   0.02007  0.00007  
5   0.06849  0.03151  
6   0.02524  0.00524  
7   0.12540  0.05460  
8   0.05791  0.03791  
9   0.08392  0.03608  
10  0.02930  0.01070  
11  0.00640  0.05360  
12  0.06857  0.04857  
13  0.07103  0.03103  
14  0.10141  0.09859  
15  0.06076  0.00076  
16  0.06811  0.04811

In [19]:
data[['a1','a2']].max()

a1    0.09859
a2    0.12540
dtype: float64

*Kolmogorov-smirnov hitung = 0.12540*

## Menghitung nilai Kolmogorov-smirnov table

**Jumlah data = 50, nilai alfa = 0.05**

*rumus untuk n > 35 dengan alfa = 0.05* adalah $1.36 / sqrt(n)$

In [20]:
import math
ks_table = 1.36/math.sqrt(50)
print("Kolmogorov-smirnov table = ",ks_table)

Kolmogorov-smirnov table =  0.19233304448274094


H0 **ditolak** jika *nilai kolmogorov-smirnov hitung >= nilai kolmogorov-smirnov table* <br>
H0 **diterima** jika *nilai kolmogorov-smirnov hitung < nilai kolmogorov-smirnov table*

**Hipotesis :** <br><li>*H0 = Data berdistribusi normal*</li><br><li>*H1 = Data tidak berdistribusi normal*</li>

*Nilai Kolmogorov-smirnov hitung >= Nilai Kolmogorov table*<br>**0.12540 < 0.1923 <br> Maka H0 diterima**

# Kesimpulannya, data tersebut berdistribusi normal